### PR TITLE
220204 netlog kernel

### PR DESCRIPTION
**Pull request by pillip8282**
This PR contains following features. 
1) ctrlog (CLI program to configure network log)
2) network kernel side log module
3) lwIP run-time log module
4) communication channel to post user request(set log config) to kernel side.
5) TC to verify lwIP runtime log and kernel log

Control Log(program name "ctrlog") can configure log for following modules
1) wi-fi manager
2) network library
3) network manager
4) lwIP

### CODE REVIEW

File: apps/examples/control_log/ctrlog.c [hunk 1]
```diff
@@ -0,0 +1,391 @@
+/****************************************************************************
+ *
+ * Copyright 2021 Samsung Electronics All Rights Reserved.
```
Comment (dineshdas-samsung) [inline apps/examples/control_log/ctrlog.c:3]: Can we please change the year to 2022?

File: apps/examples/control_log/ctrlog.c [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+	opt->color = _get_color(argv[0] + strlen(CMD_COLOR));
+	opt->mod = _get_mod(argv[1] + strlen(CMD_MOD));
+	opt->func = _handle_color;
+	printf("[pkbuild] color %d %d\t%s:%d\n", opt->color, opt->mod, __FUNCTION__, __LINE__);
+
+	return 0;
+}
+
+static int _parse_reset(void *arg, int argc, char **argv) {
+  printf("[pkbuild]  \t%s:%d\n",  __FUNCTION__, __LINE__);
```
Comment (dineshdas-samsung) [inline apps/examples/control_log/ctrlog.c:341]: Can we align the code please?

Comment (pillip8282) [reply apps/examples/control_log/ctrlog.c]: Thanks for review. I'll modify it

File: apps/examples/control_log/ctrlog.c [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+	return NL_COLOR_DEFAULT;
+}
+
+static netlog_module_e _get_mod(char *mod_str)
+{
+	printf("[pkbuild] %s \t%s:%d\n", mod_str, __FUNCTION__, __LINE__);
+	int size = sizeof(g_mod_table_str) / sizeof(char *);
+	for (int i = 0; i < size; i++) {
+		/* lwIP has additional submodule. So instead of comparing null
+     * chracter at the end of string, it won't compare it */
```
Comment (abhishek-samsung) [inline apps/examples/control_log/ctrlog.c]: Please change 'chracter' to 'character '

File: apps/examples/control_log/Kconfig [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#
+# For a description of the syntax of this configuration file,
+# see kconfig-language at https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt
+#
+
+config EXAMPLES_CTRLOG
+	bool "network logging tool"
+  depends on NET_LOGGER
```
Comment (jeongchanKim) [inline apps/examples/control_log/Kconfig:8]: Please check the indentation.

File: apps/examples/control_log/ctrlog.c [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+static nl_options g_opt_table[] = {
+		NL_OPT_ENABLE,
+		NL_OPT_DISABLE,
+};
+
+static uint32_t _get_lwip_mod(char *mod_str)
+{
+	int size = sizeof(g_lwip_mod_table_str) / sizeof(char *);
+	for (int i = 0; i < size; i++) {
+		if ((strlen(mod_str) == strlen(g_lwip_mod_table_str[i])) && strncmp(mod_str, g_lwip_mod_table_str[i], strlen(mod_str) + 1) == 0) {
```
Comment (jeongchanKim) [inline apps/examples/control_log/ctrlog.c:148]: How about declaring local variable for `strlen(mod_str)`?
Now, we should calculate the length everytime, even for `strncmp`.
If we declare one local variable and use it, we can optimize the code.

File: apps/examples/control_log/ctrlog.c [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+	}
+	return NL_COLOR_DEFAULT;
+}
+
+static netlog_module_e _get_mod(char *mod_str)
+{
+	int size = sizeof(g_mod_table_str) / sizeof(char *);
+	for (int i = 0; i < size; i++) {
+		/* lwIP has additional submodule. So instead of comparing null
+		 * chracter at the end of string, it won't compare it */
```
Comment (jeongchanKim) [inline apps/examples/control_log/ctrlog.c:182]: Please fix the typo, `chracter`.

File: apps/examples/control_log/ctrlog_usage.h [hunk 7]
```diff
@@ -0,0 +1,44 @@
+/****************************************************************************
+ *
+ * Copyright 2021 Samsung Electronics All Rights Reserved.
```
Comment (jeongchanKim) [inline apps/examples/control_log/ctrlog_usage.h:3]: Please change to 2022.

File: apps/examples/control_log/ctrlog_utils.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
+/****************************************************************************
+ *
+ * Copyright 2021 Samsung Electronics All Rights Reserved.
```
Comment (jeongchanKim) [inline apps/examples/control_log/ctrlog_utils.h:3]: Please change to 2022.

File: os/include/tinyara/lwnl/lwnl.h [hunk 9]
```diff
@@ -73,6 +73,7 @@ typedef struct {
 
 typedef enum {
 	LWNL_REQ_COMMON_GETADDRINFO,
+  LWNL_REQ_COMMON_NETLOG,
```
Comment (jeongchanKim) [inline os/include/tinyara/lwnl/lwnl.h:76]: Please check the indentation.

File: os/include/tinyara/net/netlog_lwnl.h [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+#pragma once
+
+#include <stdint.h>
+#include "netlog_util.h"
+
+typedef enum {
+  NL_OPS_TIMER,
```
Comment (jeongchanKim) [inline os/include/tinyara/net/netlog_lwnl.h:24]: Please use tab instead of space for indentation.

File: apps/examples/control_log/ctrlog.c (hunk 3 above)
Comment (pillip8282) [reply apps/examples/control_log/ctrlog.c]: Thannks I'll modify it

File: apps/examples/control_log/Kconfig (hunk 4 above)
Comment (pillip8282) [reply apps/examples/control_log/Kconfig:8]: Thannks I'll modify it

File: apps/examples/control_log/ctrlog.c (hunk 5 above)
Comment (pillip8282) [reply apps/examples/control_log/ctrlog.c:148]: Thanks for your opinion. I also considered the case you mentioned. But I thought that the function wouldn't be called frequently. Also strlen is called twice. So I thought that it's acceptable.

File: apps/examples/control_log/ctrlog.c (hunk 6 above)
Comment (pillip8282) [reply apps/examples/control_log/ctrlog.c:182]: Thannks I'll modify it

File: apps/examples/control_log/ctrlog_usage.h (hunk 7 above)
Comment (pillip8282) [reply apps/examples/control_log/ctrlog_usage.h:3]: Thannks I'll modify it. Actually it's written end of 2021 :)

File: apps/examples/control_log/ctrlog_utils.h (hunk 8 above)
Comment (pillip8282) [reply apps/examples/control_log/ctrlog_utils.h:3]: Thannks I'll modify it

File: os/include/tinyara/lwnl/lwnl.h (hunk 9 above)
Comment (pillip8282) [reply os/include/tinyara/lwnl/lwnl.h:76]: Thannks I'll modify it

File: os/include/tinyara/net/netlog_lwnl.h (hunk 10 above)
Comment (pillip8282) [reply os/include/tinyara/net/netlog_lwnl.h:24]: Thannks I'll modify it

File: .clang-format [hunk 11]
```diff
@@ -14,10 +14,10 @@ BraceWrapping:
   AfterNamespace: false
   BeforeElse: false #[M15]
 IndentCaseLabels: false #[M09]
-IndentWidth:     4 #[M08]
+IndentWidth:     2 #[M08]
```
Comment (sunghan-chang) [inline .clang-format:17]: TizenRT coding rule is 4, not 2. Please revert this and below.

File: apps/examples/control_log/Kconfig [hunk 12]
```diff
@@ -0,0 +1,16 @@
+
+#
+# For a description of the syntax of this configuration file,
+# see kconfig-language at https://www.kernel.org/doc/Documentation/kbuild/kconfig-language.txt
+#
+
+config EXAMPLES_CTRLOG
+	bool "network logging tool"
```
Comment (sunghan-chang) [inline apps/examples/control_log/Kconfig:8]: Is this a tool? Then how about moving this into apps/system, not example?

File: apps/examples/control_log/ctrlog.c [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+};
+
+static char *g_opt_table_str[] = {
+		"on",
+		"off",
+};
+
+static nl_options g_opt_table[] = {
+		NL_OPT_ENABLE,
+		NL_OPT_DISABLE,
```
Comment (sunghan-chang) [inline apps/examples/control_log/ctrlog.c:141]: Do you intend two tabs?

File: .clang-format (hunk 11 above)
Comment (pillip8282) [reply .clang-format:17]: TizenRT coding rule uses tab not space. So I think it does not violate the rule.

File: apps/examples/control_log/Kconfig (hunk 12 above)
Comment (pillip8282) [reply apps/examples/control_log/Kconfig:8]: Thanks for your suggestion. I'll move it to apps/system when it is stable and supports more network module later.

**Pull request closed without merging**